File: src/advent/2017/day4.clj
(ns advent.2017.day4
  "Advent of Code 2017, day 4: Unique passphrases"
  (:require [clojure.string :refer [join]]
            [advent.helpers :as h]))

(def puzzle-input (h/slurp-resource "2017/day4.txt" h/slurp-word-lines))

(defn- valid-passphrase? [mapper passphrase]
  (= (count passphrase) (count (distinct (map mapper passphrase)))))

(defn- valid-passphrases [mapper passphrases]
  (count (filter #(valid-passphrase? mapper %) passphrases)))

(defn puzzle1
  "Count lines with unique words"
  [passphrases]
  (valid-passphrases identity passphrases))

(defn puzzle2
  "Count lines wit no anagrams"
  [passphrases]
  (valid-passphrases #(join "" (map str (sort %))) passphrases))

File: src/haskell/2017/Day4.hs
{-|
Module      : Day4
Description : Day 4: High-Entropy Passphrases

<https://adventofcode.com/2017/day/4>
-}

module Day4 where

import Advent

import Data.List (sort)
import Data.Set (Set)
import qualified Data.Set as Set

validPassphrase :: [String] -> Bool
validPassphrase s =
  go s Set.empty where
    go (x:xs) s
      | Set.member x s = False
      | otherwise = go xs $ Set.insert x s
    go _ _ = True

main = do
    input <- parsedInput (2017, 4) (map words . lines)
    print $ length $ filter validPassphrase input
    print $ length $ filter validPassphrase $ map (map sort) input
